agent/src/beerocks/slave/tasks/task.h

```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TASK_H_
+#define _TASK_H_
+
+#include <beerocks/tlvf/beerocks_header.h>
+#include <tlvf/CmduMessageRx.h>
+
+namespace beerocks {
+
+class Task {
+public:
+    Task() {}
+    virtual ~Task() {}
+
+    /**
+     * @brief Enum which defines all tasks type.
+     * 
+     * Each task must have an entry on the enum eType, so it will be possible to send the task an
+     * event. 
+     */
+    enum eType : uint8_t {
+        TOPOLOGY,
+    };
```

Comment:
I thought about defining it in a separate class or file, but then I thought it would be better if the user will have to write `Task::eType::TASK_NAME`. It put the task name in the context of the Task class which I think is more correct.
Also, IMO it is ok to define the type on Task class because it defines what kind of tasks could be derived from it, just like pure virtual function force the derived class to implement them.

As for the `get_task_type()`. Actually that what I did it first. But when I used it looked a little bit weird:
```cpp
    auto topology_task = std::make_shared<TopologyTask>(*this, cmdu_tx);
    
    // Adding the enum to the constructor seems redundant because I already defined task of type "TopologyTask", so
    // it look weird to state it again as a constructor argument.
    m_task_pool.add_task(Task::eType::TOPOLOGY, topology_task);

    // This looks better IMO. The topology task "know" which task type is it. No need to tell it.
    m_task_pool.add_task(topology_task);
```

I don't think there is a right or wrong here.

Reply:
That's not what I meant. The task type should be an argument to the constructor of the base Task class, so that when you are implementing a new task, you'll pass it's type to the base class:
```c++
class TopologyTask : public Task
{
    TopologyTask() : Task(eTaskType::Topology) {}
    ...
};
```

`add_task` should only get the task instance as it'll be able to read the type by calling `get_task_type()`.

And no, it's not OK to modify the base class every time you add a new task.
It should be an interface (as there's no implementation here right now).

Reply:
It is not OK to know in the base class the type of the derived class. period.
It breaks the essence of interface.
Once you know the type of the derived there is no more meaning to have a base class.
Adding an enum to define the type of the concrete instance is to know its type.
Instead, we should work with the type itself, and mass around with numbers-that-represent-the-concrete-type (enum).
enums to indicate types are just another layer of complexity.

Reply:
Saying that it's not ok is a bit of an overstatement. It's perfectly fine.
At some point you have to figure out the type of the class you're working with. 
Polymorphism is not only about implementing interfaces but also about extending the interface with additional functionality which is not necessarily available in the interface you're implementing.
It can be achieved in several ways, and storing the type in the base class is one of them.
I guess you could also use `dynamic_cast` but in some (most?) implementations it relies on RTTI, which can be disabled in embedded systems.

Reply:
I moved the enum type outside of the class, but still on the same file, and set the infrastructure to what Vitali has suggested.